File: VSCode_Python/541.reverse-string-ii.py
#
# @lc app=leetcode id=541 lang=python3
#
# [541] Reverse String II
#

# @lc code=start
class Solution:
    def reverseStr(self, s: str, k: int) -> str:
        characters = [c for c in s]
        n = len(s)
        i, j = 0, k-1
        while j < n:
            start = i
            while i < j:
                characters[i], characters[j] = characters[j], characters[i]
                i += 1
                j -= 1
            i = start + 2 * k
            j = i + k-1
        if i < n:
            j = n-1
            while i < j:
                characters[i], characters[j] = characters[j], characters[i]
                i += 1
                j -= 1
        return ''.join(characters)


# @lc code=end

File: VSCode_CPP/541.reverse-string-ii.cpp
/*
 * @lc app=leetcode id=541 lang=cpp
 *
 * [541] Reverse String II
 */

// @lc code=start
class Solution {
public:
    string reverseStr(string s, int k) {
        string ans;
        int n = s.size();
        int cur=0, reverse_end=0, period_end=0;
        int i;
        while (cur < n) {
            reverse_end = min(cur+k-1, n-1);
            period_end = min(cur+k*2-1, n-1);
            for(i=reverse_end; i>=cur; i--) ans += s[i];
            for(i=reverse_end+1; i<= period_end; i++) ans += s[i];
            cur += 2*k;
        }
        return ans;
    }
};
// @lc code=end

